File: common/src/main/java/com/nivasafinance/common/interceptor/UserContextInterceptor.java
package com.nivasafinance.common.interceptor;

import com.nivasafinance.common.constants.AuthConstants;
import com.nivasafinance.common.context.UserContext;
import jakarta.servlet.http.HttpServletRequest;
import jakarta.servlet.http.HttpServletResponse;
import lombok.RequiredArgsConstructor;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;
import org.springframework.stereotype.Component;
import org.springframework.web.servlet.HandlerInterceptor;

/**
 * Interceptor to extract username from request header and set it in UserContext.
 * The API Gateway forwards the username in the X-Username header after JWT validation.
 * When X-Username is absent, falls back to resolved username from JWT (email/phone) via request attribute.
 */
@Component
@RequiredArgsConstructor
public class UserContextInterceptor implements HandlerInterceptor {

    private static final Logger logger = LoggerFactory.getLogger(UserContextInterceptor.class);
    private static final String USERNAME_HEADER = "X-Username";
    private static final String SYSTEM_USERNAME = "system";

    @Override
    public boolean preHandle(
            HttpServletRequest request,
            HttpServletResponse response,
            Object handler) {

        String username = request.getHeader(USERNAME_HEADER);
        if (username == null || username.isBlank()) {
            Object resolved = request.getAttribute(AuthConstants.RESOLVED_USERNAME_ATTRIBUTE);
            if (resolved instanceof String s && !s.isBlank()) {
                username = s;
            }
        }
        if (username == null || username.isBlank()) {
            logger.error("Missing mandatory {} header in request: {} {}",
                    USERNAME_HEADER, request.getMethod(), request.getRequestURI());
             username = SYSTEM_USERNAME;
        }

     //   userStatusValidator.validateActiveUser(username); //TODO

        UserContext.setUsername(username);
        logger.info("UserContext set for username: {}", username);

        return true;
    }

    @Override
    public void afterCompletion(
            HttpServletRequest request,
            HttpServletResponse response,
            Object handler,
            Exception ex) {
        // Clear ThreadLocal to avoid memory leaks
        UserContext.clear();
        logger.info("UserContext cleared after request completion");
    }
}

File: common/src/main/java/base/interceptor/UserContextInterceptor.kt
package base.interceptor

import base.context.UserContext
import base.model.UserInfo
import com.auth0.jwt.JWT
import com.auth0.jwt.exceptions.JWTDecodeException
import jakarta.servlet.http.HttpServletRequest
import jakarta.servlet.http.HttpServletResponse
import org.slf4j.LoggerFactory
import org.springframework.stereotype.Component
import org.springframework.web.servlet.HandlerInterceptor

@Component
class UserContextInterceptor : HandlerInterceptor {

    private val logger = LoggerFactory.getLogger(UserContextInterceptor::class.java)

    companion object {
        private const val BEARER_PREFIX = "Bearer "
    }

    // Intercept the request before it reaches the controller
    override fun preHandle(
        request: HttpServletRequest,
        response: HttpServletResponse,
        handler: Any
    ): Boolean {
        val authHeader = request.getHeader("Authorization")

        if (authHeader.isNullOrBlank() || !authHeader.startsWith(BEARER_PREFIX)) {
            logger.warn("Missing or invalid Authorization header")
            response.status = HttpServletResponse.SC_UNAUTHORIZED
            response.writer.write("Unauthorized: Missing or invalid token")
            return false
        }

        val token = authHeader.removePrefix(BEARER_PREFIX)

        return try {
            val decodedJWT = JWT.decode(token)
            val username = decodedJWT.getClaim("username")?.asString() ?: throw RuntimeException("Invalid token: username is missing")
            val email = decodedJWT.getClaim("email")?.asString() ?: "unknown"
            val phoneNumber = decodedJWT.getClaim("phone_number")?.asString() ?: "unknown"
            val roles = decodedJWT.getClaim("cognito:groups")
                ?.asList(String::class.java)
                .orEmpty()

            val userInfo = UserInfo(username, email, phoneNumber, roles)
            UserContext.setUserInfo(userInfo)
            logger.info("UserContext set with: $userInfo")

            true
        } catch (e: JWTDecodeException) {
            logger.warn("Invalid JWT token", e)
            response.status = HttpServletResponse.SC_UNAUTHORIZED
            response.writer.write("Unauthorized: Invalid token")
            false
        }
    }

    // Optionally handle logic after the controller has executed, before view rendering
    override fun postHandle(
        request: HttpServletRequest,
        response: HttpServletResponse,
        handler: Any,
        modelAndView: org.springframework.web.servlet.ModelAndView?
    ) {
        // No logic needed for postHandle, method exists for completeness.
    }

    // Always executed after the request is fully completed
    override fun afterCompletion(
        request: HttpServletRequest,
        response: HttpServletResponse,
        handler: Any,
        ex: Exception?
    ) {
        // Clear ThreadLocal to avoid memory leaks
        UserContext.clear()
        logger.info("UserContext cleared after request completion.")
    }
}
